### PR TITLE
feat: remove elf global and replace with eventListener

### PR DESCRIPTION
**Pull request by Theeraphat-Sorasetsakul**
## Description

Remove global elf object and replace with event listener.

Fixes # ([ELF-495](https://jira.refinitiv.com/browse/ELF-495))

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

**Comment by Theeraphat-Sorasetsakul** [comment]
cc. @Theeraphat-Sorasetsakul

File: packages/theme-compiler/src/helpers.js [hunk 1]
```diff
@@ -7,8 +7,13 @@ const autoprefixer = require('postcss')().use(require('autoprefixer'));
...
 const path = require('path');
 
-const wrap = (name, style) => `elf.${name.indexOf('-') > 0 ? 'custom'
-: 'native'}Styles.define('${name}', '${style.replace(/'/g, '\\\'')}');\n`;
+const wrap = (name, style, isEvent) => {
+  if(isEvent) {
+    return `dispatchEvent(new CustomEvent('ef.${name.indexOf('-') > 0 ? 'custom'
+    : 'native'}Styles.define', { detail: { name: '${name}', styles: '${style.replace(/'/g, '\\\'')}' }}));\n`;
+  }
+  return `elf.${name.indexOf('-') > 0 ? 'custom' : 'native'}Styles.define('${name}', '${style.replace(/'/g, '\\\'')}');\n`;
```
Comment (Sakchai-Refinitiv) [inline packages/theme-compiler/src/helpers.js]: The condition for checking the custom/native styles should be moved out of the template literal

File: packages/theme-compiler/src/helpers.js [hunk 2]
```diff
@@ -7,8 +7,13 @@ const autoprefixer = require('postcss')().use(require('autoprefixer'));
 const clean = new (require('clean-css'))({ returnPromise: true, level: '2' });
 const path = require('path');
 
-const wrap = (name, style) => `elf.${name.indexOf('-') > 0 ? 'custom'
-: 'native'}Styles.define('${name}', '${style.replace(/'/g, '\\\'')}');\n`;
+const wrap = (name, style, isEvent) => {
```
Comment (Sakchai-Refinitiv) [inline packages/theme-compiler/src/helpers.js:17]: Should have JSDoc

File: packages/theme-compiler/src/helpers.js [hunk 3]
```diff
@@ -41,13 +46,13 @@ const generateLessOptions = (entrypoint, filename, variables) => ({
   ]
 });
 
-const generateJsInfo = (name, css, dependencies) => {
+const generateJsInfo = (name, css, dependencies, variables) => {
```
Comment (Sakchai-Refinitiv) [inline packages/theme-compiler/src/helpers.js:71]: Should have JSDoc

File: packages/theme-compiler/src/helpers.js [hunk 4]
```diff
@@ -59,13 +64,13 @@ const getElementNameFromLess = (filename) => {
   return path.basename(filename).replace(/\.less$/, '');
 };
 
-const generateOutput = (filename, output) => {
+const generateOutput = (filename, output, variables) => {
```
Comment (Sakchai-Refinitiv) [inline packages/theme-compiler/src/helpers.js:96]: Should have JSDoc

File: packages/core/src/index.ts [hunk 5]
```diff
@@ -74,10 +74,19 @@ import { CustomStyleRegistry } from './registries/CustomStyleRegistry.js';
...
+global.addEventListener('ef.customStyles.define', (event) => {
+  const { name, styles } = (event as EfStylesDefine).detail;
+  CustomStyleRegistry.define(name, styles);
+});
 
-Object.freeze(global.elf);
+global.addEventListener('ef.nativeStyles.define', (event) => {
+  const { name, styles } = (event as EfStylesDefine).detail;
+  NativeStyleRegistry.define(name, styles);
+});
```
Comment (TremayneChrist) [inline packages/core/src/index.ts]: Don't worry about interface, just cast to string inside of the event listener. This section is temporary and will be removed soon.

File: packages/elemental-theme/package.json [hunk 6]
```diff
@@ -15,8 +15,8 @@
     "!.*"
   ],
   "scripts": {
-    "build:light": "theme-compiler light --variant=light",
-    "build:dark": "theme-compiler dark --variant=dark",
+    "build:light": "theme-compiler light --variant=light --styles=event",
```
Comment (TremayneChrist) [inline packages/elemental-theme/package.json]: Maybe use `--registration` instead of `--styles`.

File: packages/core/src/index.ts (hunk 5 above)
Comment (Theeraphat-Sorasetsakul) [reply packages/core/src/index.ts]: I think this cast cannot turn to be string. The purpose of this cast uses for eslint.

File: packages/elemental-theme/package.json (hunk 6 above)
Comment (Theeraphat-Sorasetsakul) [reply packages/elemental-theme/package.json]: Updated.

File: packages/theme-compiler/src/helpers.js (hunk 4 above)
Comment (Theeraphat-Sorasetsakul) [reply packages/theme-compiler/src/helpers.js:96]: Updated.

File: packages/theme-compiler/src/helpers.js (hunk 3 above)
Comment (Theeraphat-Sorasetsakul) [reply packages/theme-compiler/src/helpers.js:71]: Updated.

File: packages/theme-compiler/src/helpers.js (hunk 2 above)
Comment (Theeraphat-Sorasetsakul) [reply packages/theme-compiler/src/helpers.js:17]: Updated.

File: packages/theme-compiler/src/helpers.js (hunk 1 above)
Comment (Theeraphat-Sorasetsakul) [reply packages/theme-compiler/src/helpers.js]: Updated.

File: packages/core/src/index.ts (hunk 5 above)
Comment (Theeraphat-Sorasetsakul) [reply packages/core/src/index.ts]: Updated.

File: packages/theme-compiler/src/helpers.js (hunk 1 above)
Comment (Sakchai-Refinitiv) [reply packages/theme-compiler/src/helpers.js]: @Theeraphat-Sorasetsakul, You should use the `eventName` on line 19

Comment (Theeraphat-Sorasetsakul) [reply packages/theme-compiler/src/helpers.js]: Updated.

**Review by Sakchai-Refinitiv — DISMISSED** [review]

**Review by Sakchai-Refinitiv — APPROVED** [review]

**Pull request merged**